### PR TITLE
Changed the metrics logging logic

### PR DESCRIPTION
**Pull request by DomInvivo**
@zhiyil-graphcore and @luis-mueller , I merged your logics of how to reduce training time by not returning large but unused tensors. This is on top of PR #391

### CODE REVIEW

File: graphium/trainer/predictor.py
Comment (zhiyil-graphcore) [inline graphium/trainer/predictor.py:452]: Does this mean if `(batch_idx + 1) % self.metrics_every_n_train_steps != 0`, `self.is_log_train_metrics` is true and we only log the metrics in `concatenated_metrics_logs` which is global loss, `batch_tput` stuff. When `(batch_idx + 1) % self.metrics_every_n_train_steps == 0`, we log `preds`, `targets` as well as individual losses? This sounds good to me.

File: graphium/trainer/predictor.py
Comment (zhiyil-graphcore) [inline graphium/trainer/predictor.py:493]: I tested the debug config with `metrics_every_n_train_steps: 30` but somehow the individual task losses were not reported (see [here](https://wandb.ai/multitask-gnn/neurips2023_large_data_gcn_debug/runs/hss92bbw?workspace=user-zhiyi-li1230)), just wondering if there's anything missing in my setting?

File: graphium/trainer/predictor.py
Comment (zhiyil-graphcore) [inline graphium/trainer/predictor.py:483]: Also got this error when testing `config_large_gcn.yaml`:
```
`  File "/nethome/zhiyil/graphium/graphium/ipu/ipu_wrapper.py", line 105, in on_train_batch_end
    super().on_train_batch_end(outputs, batch, batch_idx)
  File "/nethome/zhiyil/graphium/graphium/trainer/predictor.py", line 481, in on_train_batch_end
    targets=outputs["targets"],
KeyError: 'targets'
[10:29:23.668] [poptorch::python] [critical] KeyError: 'targets'`
```

File: graphium/trainer/predictor.py
Comment (DomInvivo) [reply graphium/trainer/predictor.py:452]: Yes exactly!

File: graphium/trainer/predictor.py
Comment (DomInvivo) [reply graphium/trainer/predictor.py:483]: Yeah I forgot to mention that I didn't test the code at all. It was just easier to do a wick draft than to explain in the other PR what I had in mind. 

If it's not too much to ask, can you double check everything before merging?

File: graphium/trainer/predictor.py
Comment (DomInvivo) [reply graphium/trainer/predictor.py:493]: See my comment below, this is a draft I have made and could have mistakes. Feel free to edit this branch before merging

File: graphium/trainer/predictor.py
Comment (DomInvivo) [reply graphium/trainer/predictor.py:452]: I changed `batch_idx + 1` --> `batch_idx` because I think it's a good idea to log the first batch.

File: graphium/trainer/predictor.py
Comment (DomInvivo) [reply graphium/trainer/predictor.py]: I think I fixed that in the latest commit, I moved that part before the `if self.skip_log_train_metrics`

Comment (zhiyil-graphcore) [reply graphium/trainer/predictor.py]: I also moved the `concatenated_metrics_logs[key] = concatenated_metrics_logs[key].mean()` part up so that we get averaged loss for the individual losses.

File: graphium/trainer/predictor.py
Comment (zhiyil-graphcore) [reply graphium/trainer/predictor.py:492]: I tested with debug config, seems to be fine.

Comment (zhiyil-graphcore) [reply graphium/trainer/predictor.py:492]: The test on `config_large_gcn.yaml` shows that it takes 20 minutes per epoch, which is 10 minutes longer than before...

Comment (DomInvivo) [reply graphium/trainer/predictor.py:492]: Perhaps it's the `gradient_norm` that's taking a long time?

Comment (DomInvivo) [reply graphium/trainer/predictor.py:492]: Or perhaps the logic is wrong and things gets logged more often than once every 1000 steps?

Comment (zhiyil-graphcore) [reply graphium/trainer/predictor.py:492]: I'll look into that, currently I'm not logging any metrics only the losses. Are "preds" and "targets" only used for metric logging here?

BTW, I found a way to filter 0 losses from different replica or gradient accumulation. `oncatenated_metrics_logs[key] = concatenated_metrics_logs[key].mean()` is actually averaging on r`eplica * gradien_accu * device_iter`. In `ipu_wrapper.py`, there is another averaging `outputs["loss"] = outputs["loss"].mean()`. We might just need to filter out the 0s before doing the averaging. From another point of view, as the total loss is the sum of the 5 task losses, it would not end up with zero anyway I think? We may just need to filter on individual loss for the reporting purpose.

Comment (zhiyil-graphcore) [reply graphium/trainer/predictor.py:492]: > Or perhaps the logic is wrong and things gets logged more often than once every 1000 steps?

maybe it's not that bad after all, it's taking around 22 hours for 100 epochs, compared to 19 hours previously, it is quite ok I think.

**Review by zhiyil-graphcore — APPROVED** [review]

File: graphium/trainer/predictor.py
Comment (luis-mueller) [inline graphium/trainer/predictor.py:499]: It seems that `outputs` is never used anywhere but also that `metrics_logs` never makes it to `concatenated_metrics_logs`, hence the extended metrics don't appear on `wandb`. @zhiyil-graphcore can you confirm this on your side?

My fix would be to simply remove the `outputs.update` call and replace with `concatenate_metrics_logs.update(metrics_logs)`. This works on my side at least to log the extended metrics every `n` steps but the main metrics at every step.

Comment (luis-mueller) [reply graphium/trainer/predictor.py:499]: If you agree I will push my change and approve.

Comment (zhiyil-graphcore) [reply graphium/trainer/predictor.py:499]: I turned off most metrics during training so didn't catch that point. Feel free to push the changes and I can verify the throughput on IPU.

Comment (DomInvivo) [reply graphium/trainer/predictor.py:499]: I have made some changes, so there might be some conflicts, please beware

File: graphium/trainer/predictor.py
Comment (DomInvivo) [reply graphium/trainer/predictor.py:492]: I removed the throughput computation and the `get_gradient_norm` if `self.skip_log_train_metrics is True`, so things should be a bit faster now

File: graphium/trainer/losses.py
Comment (luis-mueller) [inline graphium/trainer/losses.py:66]: These changes were necessary to enable gradient computation on my side (the previous, in-place modification caused this tedious error: `one of the variables needed for gradient computation has been modified by an in-place operation`). Not sure why the tests aren't catching this or whether this is just an issue on my side. @DomInvivo your call whether we apply this change, otherwise I can also remove it. Will definitely not be as fast and memory-friendly as the in-place but I also cannot estimate how impactful it really is to change to out-of-place (essentially creates a copy but since I am assigning it to the same variable, the garbage collector should remove the old tensor)

**Review by luis-mueller — APPROVED** [review]
Worked on GPU side, see comments on in-place operations in loss and my fix to actually log the metrics. LGTM

**Pull request merged**
